src/python/pants/engine/target.py

```diff
@@ -703,7 +724,63 @@ def compute_value(
                     "either a single string or an iterable of strings (e.g. a list of strings)"
                 ),
             )
-        return tuple(sorted(str_list))
+        return tuple(str_list)
+
+
+class DictStringToStringField(PrimitiveField, metaclass=ABCMeta):
+    value: Optional[FrozenDict[str, str]]
+    default: ClassVar[Optional[FrozenDict[str, str]]] = None
+
+    @classmethod
+    def compute_value(
+        cls, raw_value: Optional[Dict[str, str]], *, address: Address
+    ) -> Optional[FrozenDict[str, str]]:
+        value_or_default = super().compute_value(raw_value, address=address)
+        if value_or_default is None:
+            return None
+        invalid_type_exception = InvalidFieldTypeException(
+            address, cls.alias, raw_value, expected_type="a dictionary of string -> string"
+        )
+        if not isinstance(value_or_default, dict):
+            raise invalid_type_exception
+        if not all(isinstance(k, str) and isinstance(v, str) for k, v in value_or_default.items()):
+            raise invalid_type_exception
+        return FrozenDict(value_or_default)
+
+
+class DictStringToStringSequenceField(PrimitiveField, metaclass=ABCMeta):
+    value: Optional[FrozenDict[str, Tuple[str, ...]]]
+    default: ClassVar[Optional[FrozenDict[str, Tuple[str, ...]]]] = None
+
+    @classmethod
+    def compute_value(
+        cls, raw_value: Optional[Dict[str, Iterable[str]]], *, address: Address
```

Comment:
I'm not a super big fan of flexible constructors, especially if you're going to transform them into tuples later. The contract of `Iterable` is "I can traverse it somehow", whereas [a part of] the contract of `Tuple` is  "I can traverse it in the same order every time". I really wish there was something like `OrderedIterable`, which would contain things like `List`, `Tuple`, `OrderedSet` and `FrozenOrderedSet`, but not `set` or `dict`. That way, the API change of `dicts` being ordered now would also be reflected in the type system, because it would now also be an `OrderedIterable`.

Reply:
> I really wish there was something like OrderedIterable

There is. `typing.Sequence`. 

But we need `typing.Iterable` unfortunately because we need to allow users to keep using sets in their BUILD files. Daniel went out of his way a few months ago iirc to fix up support for sets. 

It'd be nice to remove support for sets, but we'd need to do that in a followup.